First draft of 1.35 announcement (#391)
rust-lang/blog.rust-lang.org: posts/2019-05-23-Rust-1.35.0.md

---
layout: post
title: "Announcing Rust 1.35.0"
author: The Rust Release Team
release: true
---

The Rust team is happy to announce a new version of Rust, 1.35.0. Rust is a
programming language that is empowering everyone to build reliable and efficient software.

If you have a previous version of Rust installed via rustup, getting Rust 1.35.0 is as easy as:

```console
$ rustup update stable
```

If you don't have it already, you can [get `rustup`][install] from the appropriate page on our website,
and check out the [detailed release notes for 1.35.0][notes] on GitHub.

[install]: https://www.rust-lang.org/install.html
[notes]: https://github.com/rust-lang/rust/blob/master/RELEASES.md#version-1350-2019-05-23

## What's in 1.35.0 stable

The highlight of this release is the implementation of the `FnOnce`, `FnMut`,

> and `Fn` closure traits for `Box<dyn FnOnce>`, `Box<dyn FnMut>`, and `Box<dyn Fn>` respectively.

Reviewer: Is it worth clarifying at some point that it's for `Box<F>` where `F: Fn* + ?Sized`?

Author: @Manishearth ^---

Reviewer 2: Oh, yeah, can someone do that?

Additionally, closures may now be coerced to unsafe function pointers.
The `dbg!` macro introduced in [Rust 1.32.0] can now also be called without arguments.
Moreover, there were a number of standard library stabilizations.
Read on for a few highlights, or see the [detailed release notes][notes] for additional information.

> ### `Fn*` closure traits implemented for `Box<dyn Fn*>`

Author: r? @cramertj

Reviewer 3: LGTM!

Author: Cool. :)

[fn-pr]: https://github.com/rust-lang/rust/pull/55431
[`FnBox`]: https://doc.rust-lang.org/1.34.0/std/boxed/trait.FnBox.html
[unsized locals]: https://doc.rust-lang.org/nightly/unstable-book/language-features/unsized-locals.html

In Rust 1.35.0, the `FnOnce`, `FnMut`, and the `Fn` traits [are now implemented][fn-pr] for `Box<dyn FnOnce>`,
`Box<dyn FnMut>`, and `Box<dyn Fn>` respectively.

Previously, if you wanted to call the function stored in a boxed closure, you had to use [`FnBox`].

> This was because instances of `Box<dyn FnOnce>`, etc did not implement the respective `Fn*` trait.

Author: ```suggestion
This was because instances of `Box<dyn FnOnce>` and friends did not implement the respective `Fn*` traits.
```

This also meant that it was not possible to pass boxed functions to code expecting an implementor of a `Fn` trait,
and you had to create temporary closures to pass them down.

> This was ultimately due to a limitation in the compiler's ability to reason about such impls,

Author: ```suggestion
This was ultimately due to a limitation in the compiler's ability to reason about such implementations,
```

which has since been fixed with the introduction of [unsized locals].

With this release, you can now use boxed functions in places that expect items implementing a function trait.

The following code now works:

```rust
fn foo(x: Box<dyn Fn(u8) -> u8>) -> Vec<u8> {

>     vec![1,2,3,4].into_iter().map(x).collect()

Author: ```suggestion
    vec![1, 2, 3, 4].into_iter().map(x).collect()
```

This release introduces a convenience method `replace_with` on `RefCell`:

- [`RefCell::replace_with`]

With it, you can more ergonomically map and replace the current value of the cell and get back the old value as a result.

#### Hash a pointer or reference by address, not value

In this release, we have introduced:

- [`ptr::hash`]

This function takes a raw pointer and hashes it. Using `ptr::hash`,
you can avoid hashing the pointed-to value of a reference and instead hash the address.

#### Copy the contents of an `Option<&T>`

From the very beginning with Rust 1.0.0,
the methods `Option::cloned` for `Option<&T>` and `Option<&mut T>` have allowed you to clone the contents in case of `Some(_)`.
However, cloning can sometimes be an expensive operation and the methods `opt.cloned()` provided no hints to that effect.

With this release of Rust, we introduced:

- [`Option::copied`] for both `Option<&T>` and `Option<&mut T>`

The functionality of `opt.copied()` is the same as for `opt.cloned()`.
However, calling the method requires that `T: Copy`.
Using this method, you can make sure that code stops compiling should `T` no longer implements `Copy`.

> ### Changes in Clippy

Author: r? @oli-obk

[relnotes-clippy]: 

> https://github.com/rust-lang/rust-clippy/blob/master/CHANGELOG.md#rust-135-beta

Author: It does render correctly.

[`drop_bounds`]: https://rust-lang.github.io/rust-clippy/master/index.html#drop_bounds

[split_lint]: https://github.com/rust-lang/rust-clippy/pull/4101

In this release of Rust,
Clippy (a collection of lints to catch common mistakes and improve your Rust code) added a new lint [`drop_bounds`].
This lint triggers when you add a bound `T: Drop` to a generic function. For example:

```rust
fn foo<T: Drop>(x: T) {}
```

Having a bound `T: Drop` is almost always a mistake as it excludes types,
such as `u8`, which have trivial drop-glues.
Moreover, `T: Drop` does not account for types like `String` not having interesting destructor behavior directly but rather as a result of embedding types,
such as `Vec<u8>`, that do.

In addition to [`drop_bounds`],
this release of Clippy [split][split_lint] the lint`redundant_closure` into `redundant_closure` and `redundant_closure_for_method_calls`.

See the [detailed release notes for Clippy][relnotes-clippy] for more details.

> ### Changes in Cargo

Author: r? @ehuss

If you want to add any highlights please send in a patch.

Reviewer 4: I don't think there is anything worth mentioning here for this release, so I'm r+.

Author: OK; cool. resolving then =)
